### ISSUE TITLE
Insert can NOT work well when composite primary keys on MySQL

### ISSUE THREAD

**Issue by ceshihao** [volatiletech/sqlboiler#461]
My test MySQL schema is simplified like

```sql
CREATE TABLE `demo` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `date` date NOT NULL DEFAULT '0000-00-00',
  `created_at` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',
  `updated_at` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (`id`,`date`),
) ENGINE=InnoDB DEFAULT CHARSET=utf8
```

The table is taken as `un-CanLastInsertID` in the following function because of composite primary keys.
https://github.com/volatiletech/sqlboiler/blob/df8550538fefdf62f2520e74dbfad73f1bcb84a2/drivers/table.go#L44-L66

However, when populating default values for the model, it uses both 2 primary keys id = 0 (`last_insert_id` isn't used here) and date='20190123'. Nothing found and populate fails.

@aarondl 
What do you think about relaxing the condition of `CanLastInsertID()` to take composite primary keys as `true`?

**Comment by ceshihao** [comment]
I think the AUTO_INCREMENT `id` can guarantee the uniqueness in my case. I can make the second primary key `date` without default value.

e.g.
```sql
CREATE TABLE `demo` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `date` date NOT NULL,
  `created_at` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',
  `updated_at` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (`id`,`date`),
) ENGINE=InnoDB DEFAULT CHARSET=utf8
```

Actually, I can pull the right record just through AUTO_INCREMENT `id`. I add `date` as secondary primary key for some other reason.

**Comment by ceshihao** [comment]
I think the AUTO_INCREMENT `id` can guarantee the uniqueness in my case. I can make the second primary key `date` without default value.

e.g.
```sql
CREATE TABLE `demo` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `date` date NOT NULL,
  `created_at` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',
  `updated_at` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (`id`,`date`),
) ENGINE=InnoDB DEFAULT CHARSET=utf8
```

Actually, I can pull the right record just through AUTO_INCREMENT `id`. I add `date` as secondary primary key for some other reason.

**Comment by aarondl** [comment]
An auto incremented id can -always- uniquely identify a table row. I can't understand why you'd have date in there and it sounds like you don't either.

Maybe this is not a useful use case? Doesn't sound useful to me :)

**Comment by ceshihao** [comment]
`date` is used as partition key which has to be  primary/unique key.

**Comment by aarondl** [comment]
@ceshihao Sorry for not responding for a bit.

I read this documentation:
https://www.postgresql.org/docs/current/ddl-partitioning.html

And I thought about your use case (two people can order something on the same timestamp but the same person cannot) and it makes some sense to do (id, date). If you simply had a `unique date` then two customers could not place an order at the exact same time (despite that being a near impossibility, it's still a weird case).

So I think I understand better now.

But that doesn't change my response from above, if we did want to relax the constraint it would still have to have the constraints I mentioned in that:
> For a composite primary key we must have non-default values for all fields that are part of the primary key except the auto-increment one. Otherwise we can't guarantee that we'll pull the right record.

If we were to do that it'd probably be okay?

**Comment by nin2hanzo** [comment]
@aarondl 
Hello.
I am Japanese, so please forgive my poor English.

It has been a long time since this problem was reported, have you guys encountered the same problem?
In fact, I just encountered the same problem with my product and tried to debug it again.

The premise of the case I encountered is probably the same as @ceshiao who reported this issue.
What it is is that you have a composite primary key consisting of an auto-generated ID and other columns (in this case of type DATETIME). It is especially important that the "auto-generated ID" is included in the composite primary key.
The schema definition is introduced again as follows.

```sql
CREATE TABLE `demo`
(
  `id`         bigint unsigned auto_increment,
  `date`       datetime(6) not null,
  `created_at` datetime    not null,
  `updated_at` datetime    not null,
  PRIMARY KEY (`id`, `date`)
) ENGINE = InnoDB
  DEFAULT CHARSET = utf8
```

First, we noticed that the test case returns errors such as "db: unable to populate default values for demo".
However, since the date column has no default values and there are no duplicate records, we investigate and find that a record was created in the table and the INSERT actually succeeded.

Now, what is the cause of this?
As @ceshihao said, the `CanLastInsertID()` condition determines that if the table has a compound primary key, you cannot get LastInsertID even if part of the compound primary key is an auto-increment ID. However, this condition is not strict and the mysql driver can return LastInsertID if the column has the AUTO_INCREMENT attribute.

This means that the following expected operations will not be performed, and the retrieval of a successful INSERT record will fail.

https://github.com/volatiletech/sqlboiler/blob/22497a87c693908913b6f1d6572fb6541cb555dc/templates/main/15_insert.go.tpl#L147-L156

https://github.com/volatiletech/sqlboiler/blob/22497a87c693908913b6f1d6572fb6541cb555dc/templates/main/15_insert.go.tpl#L181-L189

In other words, the `o.ID` expected to be generated automatically remains 0.

Do you think it is possible to fix this problem?
